keyboards/model_f_labs/xwhatsit_util_comm.c

```diff
@@ -0,0 +1,210 @@
+            wait_ms(10);
+            bootloader_jump();
+            // we should not be reaching the following:
+            wait_ms(10);
+            response[2] = UTIL_COMM_RESPONSE_OK;
+            break;
+        case UTIL_COMM_GET_KEYSTATE:
+            response[2] = UTIL_COMM_RESPONSE_OK;
+            {
+                matrix_row_t current_matrix[MATRIX_ROWS];
+                matrix_scan_raw(current_matrix);
+                char *current_matrix_ptr = (char *)current_matrix;
+                int   offset             = 0;
+                if (sizeof(current_matrix) > 32 - 3) {
+                    offset = data[3];
+                    current_matrix_ptr += offset;
+                }
+                memcpy(&response[3], current_matrix_ptr, min(32 - 3, sizeof(current_matrix) - offset));
+            }
+            break;
+        case UTIL_COMM_GET_THRESHOLDS:
+            response[2] = UTIL_COMM_RESPONSE_OK;
+#if CAPSENSE_CAL_ENABLED
+            response[3] = CAPSENSE_CAL_BINS;
+            {
+                const uint8_t cal_bin           = data[3];
+                response[4]                     = cal_thresholds[cal_bin] & 0xff;
+                response[5]                     = (cal_thresholds[cal_bin] >> 8) & 0xff;
+                char *assigned_to_threshold_ptr = (char *)assigned_to_threshold[cal_bin];
+                int   offset                    = 0;
+                if (sizeof(assigned_to_threshold[cal_bin]) > 32 - 6) {
+                    offset = data[4];
+                    assigned_to_threshold_ptr += offset;
+                }
+                memcpy(&response[6], assigned_to_threshold_ptr, min(32 - 6, sizeof(assigned_to_threshold[cal_bin]) - offset));
+            }
+#else
+            response[3] = 0;
+            response[4] = (CAPSENSE_HARDCODED_THRESHOLD)&0xff;
+            response[5] = ((CAPSENSE_HARDCODED_THRESHOLD) >> 8) & 0xff;
+#endif
+            break;
+        case UTIL_COMM_GET_KEYBOARD_FILENAME: {
+            response[2] = UTIL_COMM_RESPONSE_OK;
+            if (data[3] >= strlen(KEYBOARD_FILENAME) + 1) {
+                response[3] = 0;
+            } else {
+                const char *substring        = KEYBOARD_FILENAME + data[3];
+                size_t      substring_length = strlen(substring) + 1;
+                if (substring_length > RAW_EPSIZE - 3) substring_length = RAW_EPSIZE - 3;
+                memcpy(&response[3], substring, substring_length);
+            }
+            break;
+        }
+        case UTIL_COMM_ERASE_EEPROM: {
+            response[2] = UTIL_COMM_RESPONSE_OK;
+            uint16_t addr;
+            for (addr = 0; addr < E2END; addr += 1) {
+                eeprom_update_byte((uint8_t *)addr, 0xff);
+            }
+            break;
+        }
+        case UTIL_COMM_GET_SIGNAL_VALUE: {
+            response[2]   = UTIL_COMM_RESPONSE_OK;
+            uint8_t col   = data[3];
+            uint8_t row   = data[4];
+            uint8_t count = data[5];
```

Comment:
```suggestion
            uint8_t count = data[5];
            if (count > (RAW_EPSIZE-3)/2) {
                count = (RAW_EPSIZE-3)/2;
            }
```
(To avoid a buffer overflow when writing to the `response` array.)

Reply:
Added the bounds checking, Thank you for your review.